autodiff/mixed/expressions/binary.hpp

```diff
@@ -0,0 +1,191 @@
+    /// so we don't need to update them.
+    virtual void update()
+    {
+        this->value() = base::left() + base::right();
+    }
+};
+
+/// @brief Binary minus expression.
+/// @tparam N Maximal derivative order of expression.
+template<std::size_t N>
+struct minus_expression final : binary_expression<N>
+{
+    using base = binary_expression<N>;
+
+    /// @brief Construct minus expression from left and right expression values.
+    /// @param l Value of left expression.
+    /// @param r Value of right expression.
+    explicit minus_expression(const double& l, const double& r) :
+        base(l - r, l, r)
+    {
+        this->template partial_derivatives<first_order>() = { 1.0, - 1.0 };
+
+        if constexpr (N > first_order)
+            this->template partial_derivatives<second_order>() = { };
+    }
+
+    /// @brief Update value of minus expression.
+    /// @details Values of partial derivatives are always same for minus, 
+    /// so we don't need to update them.
+    virtual void update()
+    {
+        this->value() = base::left() - base::right();
+    }
+};
+
+/// @brief Binary multiply expression.
+/// @tparam N Maximal derivative order of expression.
+template<std::size_t N>
+struct multiply_expression final : binary_expression<N>
+{
+    using base = binary_expression<N>;
+
+    /// @brief Construct multiply expression from left and right expression values.
+    /// @param l Value of left expression.
+    /// @param r Value of right expression.
+    explicit multiply_expression(const double& l, const double& r) : 
+        base(l * r, l, r)
+    { 
+        this->template partial_derivatives<first_order>() = { base::right(), base::left() };
+
+        if constexpr (N > first_order)
+            this->template partial_derivatives<second_order>() = { };
+    }
+
+    /// @brief Update value of multiply expression.
+    /// @details Only first order derivatives are need recalculation,
+    /// since other are always zero.
+    virtual void update()
+    {
+        this->value() = base::left() * base::right();
+        // update derivatives for multiply expression
+        this->template partial_derivatives<first_order>() = { base::right(), base::left() };
+    }
+};
+
+/// @brief Binary divide expression.
+/// @tparam N Maximal derivative order of expression.
+template<std::size_t N>
+struct divide_expression final : binary_expression<N>
+{
+    using base = binary_expression<N>;
+
+    /// @brief Construct divide expression from left and right expression values.
+    /// @param l Value of left expression.
+    /// @param r Value of right expression.
+    explicit divide_expression(const double& l, const double& r) : 
+        base(l / r, l, r)
+    { 
+        eval_derivatives();
+    }
+
+    /// @brief Update value of divide expression.
+    /// @details Only first order derivatives are need recalculation,
+    /// since other are always zero.
+    virtual void update()
```

Comment:
Wondering here if these update methods can deal with branches (e.g., if conditions)?

Reply:
Also can be introduced in new PR